javaslang/src/main/java/javaslang/collection/Iterator.java

```diff
@@ -968,7 +968,8 @@ public T getNext() {
     static <T> Iterator<T> iterate(T seed, Function<? super T, ? extends T> f) {
         Objects.requireNonNull(f, "f is null");
         return new AbstractIterator<T>() {
-            T next = seed;
+            Function<? super T, ? extends T> nextFunc = (s) -> seed;
```

Comment:
Yep, we should embrace `null` because it is Java we use. Java will have `null`s forever and so we should support this billion dollar mistake (or use another language which does not have `null`).

Please use this slightly optimized version:

```java
    static <T> Iterator<T> iterate(T seed, Function<? super T, ? extends T> f) {
        Objects.requireNonNull(f, "f is null");
        return new AbstractIterator<T>() {
            Function<? super T, ? extends T> nextFunc = s -> {
                nextFunc = f;
                return seed;
            };
            T next = null;

            @Override
            public boolean hasNext() {
                return true;
            }

            @Override
            public T getNext() {
                return next = nextFunc.apply(next);
            }
        };
    }
```

You may write

```java
            public T getNext() {
                next = nextFunc.apply(next);
                return next;
            }
```

if you are not comfortable with having a return statement and an assignment in one line. At least @paplorinc will not be comfortable with it ;-p

---
one side-note about style: we write `s -> ...` instead of `(s) -> ...` in case of single arg lambdas..